engine/src/engine.rs

```diff
@@ -723,6 +723,12 @@ impl<'env, I: IO + Copy, E: Env> Engine<'env, I, E> {
             )))
         };
 
+        if let Some(fallback_address) = silo::get_erc20_fallback_address(&self.io) {
+            if !silo::is_allow_receive_erc20_tokens(&self.io, &recipient) {
```

Comment:
But why the `erc20_fallback_address` can't be also an indicator of silo mode on?

Reply:
Ok. The question is follow. Is it possible case when the silo mode is on but the `erc20_fallback_address` is not set? Or `erc20_fallback_address` must be set always in silo mode?

Reply:
it's hard for me to unequivocally answer, but yeh, I think the `erc20_fallback_address` should be used only if the silo mode is on.
Let's make things less flexible to avoid vulnerabilities, so let's add a function `is_silo_mode_enabled()` and use it instead of the `fixed_gas_cost` everywhere.
The `is_silo_mode_enabled` could be implemented in two way just `fixed_gas_cost.is_some()` or `fixed_gas_cost.is_some() && erc20_fallback_address.is_some()`.

also, we can combine the `fixed_gas_cost` and `erc20_fallback_address` into one struct under one storage key.

Reply:
Then I would propose the following changes. Change the structure of the input parameter of the `set_fixed_gas_cost` function by adding one more field: `fallback_address`. Obviously, `set_fixed_gas_cost` should be renamed to `set_silo_params` or something. Also, we should not be able to set fixed costs without a fallback address. In such a way, presenting a fixed cost or fallback address means that silo mode is on. It allows us to eliminate `is_silo_mode_enabled`, which adds additional storage read.

Reply:
Yeh, sounds good!